radis/io/geisa_utils.py

```diff
@@ -0,0 +1,246 @@
+Essentially cloned from erwan's tools.py with adjustments to fit GEISA format
+
+@author: TranHuuNhatHuy
+"""
+
+import numpy as np
+import pandas as pd
+
+
+def parse_geisa_file(fname, columns, count=-1):
+    """Parse a file under GEISA ``par`` format. Parsing is done in binary
+    format with :py:func:`numpy.fromfile` so it's as fast as possible.
+
+    Parameters
+    ----------
+    fname: str
+        filename.
+    columns: dict
+        list of columns and their format.
+
+    Other Parameters
+    ----------------
+    count: int
+        number of lines to read. If ``-1`` reads all file.
+
+    Returns
+    -------
+    df: pandas DataFrame
+        dataframe with lines.
+
+    See Also
+    --------
+    Used in :py:func:`~radis.io.geisa.gei2df`
+    """
+
+    data = _read_geisa_file(
+        fname, columns, count=1, linereturnformat="a2"
+    )  # 'a2' allocates space to get \n or \n\r linereturn formats
+    linereturnformat = _get_linereturnformat(data, columns, fname)
+
+    # Now re-read with correct line return character
+    data = _read_geisa_file(fname, columns, count, linereturnformat)
+
+    # Return a Pandas dataframe
+    return _ndarray2df(data, columns, linereturnformat)
+
+
+def _get_linereturnformat(data, columns, fname=""):
+    """
+    Get line return character & format (size).
+    """
+    # fname just for the error message
+
+    # get format (size) of line return
+    from radis.misc.basics import to_str
+
+    linereturn = to_str(data[0][-1])
+    if to_str("\r\n") in linereturn:
+        linereturnformat = "a2"
+    elif to_str("\n") in linereturn or to_str("\r") in linereturn:
+        linereturnformat = "a1"
+    else:
+        raise ValueError(
+            "Unknown Line return format: {0}. Check that your file {1} has the GEISA format. First line : {2}".format(
+                linereturn, fname, data[0]
+            )
+        )
+
+    return linereturnformat
+
+
+def _ndarray2df(data, columns, linereturnformat):
+    """ """
+
+    newtype = [c[0] if (c[1] == str) else c[1] for c in columns.values()]
+    dtype = list(zip(list(columns.keys()), newtype)) + [
+        ("_linereturn", linereturnformat)
+    ]
+
+    data = _cast_to_dtype(data, dtype)
+
+    # %% Create dataframe
+    df = pd.DataFrame(data.tolist(), columns=list(columns.keys()) + ["_linereturn"])
+
+    # Delete dummy column than handled the line return character
+    del df["_linereturn"]
+
+    # Update format
+    for k, c in columns.items():
+        if c[1] == str:
+            df[k] = df[k].str.decode("utf-8")
+
+    # Strip whitespaces around PQR columns (due to 2 columns jumped)
+    if "branch" in df:  # (only in CDSD)
+        df["branch"] = df.branch.str.strip()
+
+    # Commence "D to E" conversion on 2nd and 20th columns, by using
+    # df.columns.str.replace() which is much faster than Python loop.
+    # Finally, typecast them to float.
+    df["int"] = df["int"].astype(str).str.replace("D", "E").astype(float)
```

Comment:
Refactor suggestion : as I understand these lines are the only difference between this file and the hitran one. If you were moving them directly in the GEISA.py in the geisa2df function you could use all the codes from tools.py  and avoid creating a new file

Reply:
Yes, initially I created a new utility file for GEISA in order to effectively separate it from HITRAN process, while dealing with those `numpy.dtype` conversion. But now seems like we don't have to go through those problematic conversions anymore (thanks to the update of GEISA columns), thus I will remove the GEISA utility file and keep these structure simple.